nupic/algorithms/KNNClassifier.py

```diff
@@ -565,27 +569,35 @@ def infer(self, inputPattern, computeScores=True,
     else:
       maxCategoryIdx = max(self._categoryList)
       inferenceResult = numpy.zeros(maxCategoryIdx+1)
-      dist = self._getDistances(inputPattern, partitionId = partitionId)
-
-      sorted = dist.argsort()
-
+      dist = self._getDistances(inputPattern, partitionId=partitionId)
       validVectorCount = len(self._categoryList) - self._categoryList.count(-1)
-      for j in sorted[:min(self.k, validVectorCount)]:
-        inferenceResult[self._categoryList[j]] += 1.0
-
-      winner = inferenceResult.argmax()
-      inferenceResult /= inferenceResult.sum()
 
+      # Loop through the indices of the nearest neighbors.
+      if self.exact:
+        exactMatches = numpy.where(dist==0.0)[0]
+        if exactMatches.any():
+          for i in exactMatches[:min(self.k, validVectorCount)]:
+            inferenceResult[self._categoryList[i]] += 1.0
+      else:
+        sorted = dist.argsort()
+        for j in sorted[:min(self.k, validVectorCount)]:
+          inferenceResult[self._categoryList[j]] += 1.0
+
+      # Prepare inference results.
+      if inferenceResult.any():
+        winner = inferenceResult.argmax()
+        inferenceResult /= inferenceResult.sum()
+      else:
+        winner = None
```

Comment:
You are changing the return semantics. Please update the comment for infer() with the meaning of None.  Also, add a test case that covers this scenario.